fil_fungible_token/src/token/mod.rs

```diff
@@ -278,8 +285,380 @@ where
+        assert_eq!(total_supply, TokenAmount::from(1_000_000));
+    }
+
+    #[test]
+    fn it_burns() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let treasury = 1;
+        let mint_amount = TokenAmount::from(1_000_000);
+        let burn_amount = TokenAmount::from(600_000);
+        token.mint(treasury, mint_amount).unwrap();
+        token.burn(treasury, treasury, burn_amount).unwrap();
+
+        // total supply decreased
+        let total_supply = token.total_supply();
+        assert_eq!(total_supply, TokenAmount::from(400_000));
+
+        // treasury balance decreased
+        let balance = token.balance_of(treasury).unwrap();
+        assert_eq!(balance, TokenAmount::from(400_000));
+    }
+
+    #[test]
+    fn it_allows_delegated_burns() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let treasury = 1;
+        let burner = 2;
+        let mint_amount = TokenAmount::from(1_000_000);
+        let approval_amount = TokenAmount::from(600_000);
+        let burn_amount = TokenAmount::from(600_000);
+
+        // mint the total amount
+        token.mint(treasury, mint_amount).unwrap();
+        // approve the burner to spend the allowance
+        token
+            .increase_allowance(treasury, burner, approval_amount)
+            .unwrap();
+        // burn the approved amount
+        token.burn(burner, treasury, burn_amount.clone()).unwrap();
+
+        // total supply decreased
+        let total_supply = token.total_supply();
+        assert_eq!(total_supply, TokenAmount::from(400_000));
+        // treasury balance decreased
+        let balance = token.balance_of(treasury).unwrap();
+        assert_eq!(balance, TokenAmount::from(400_000));
+        // burner approval decreased
+        let new_allowance = token.allowance(treasury, burner).unwrap();
+        assert_eq!(new_allowance, TokenAmount::zero());
+
+        // disallows another delegated burn as approval is zero
+        // burn the approved amount
+        token
+            .burn(burner, treasury, burn_amount)
+            .expect_err("unable to burn more than allowance");
+
+        // balances didn't change
+        let total_supply = token.total_supply();
+        assert_eq!(total_supply, TokenAmount::from(400_000));
+        let balance = token.balance_of(treasury).unwrap();
+        assert_eq!(balance, TokenAmount::from(400_000));
+        let new_allowance = token.allowance(treasury, burner).unwrap();
+        assert_eq!(new_allowance, TokenAmount::zero());
+    }
+
+    #[test]
+    fn it_cannot_burn_below_zero() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let treasury = 1;
```

Comment:
Consider some constants for addresses to move these lines out of the tests themselves.